File: python/07_functions.py
# ============================================================
# Topic:  Functions
# Summary: def, parameters, defaults, *args/**kwargs, return, lambda.
# Run:     python python/07_functions.py
# ============================================================

# ----- 1. Basic function -----
def greet(name):
    return f"Hello, {name}"
print(greet("Vetri"))        # Output: Hello, Vetri

# ----- 2. Type hints (optional; TS makes these mandatory-ish) -----
def add(a: int, b: int) -> int:
    return a + b
print(add(2, 3))             # Output: 5

# ----- 3. Default parameters -----
def power(base, exp=2):      # exp defaults to 2
    return base ** exp
print(power(5))              # Output: 25
print(power(5, 3))           # Output: 125

# ----- 4. Keyword arguments (call by name, any order) -----
def make_user(name, age):
    return f"{name}/{age}"
print(make_user(age=22, name="Vetri"))   # Output: Vetri/22

# ----- 5. *args -> variable positional args (TS: ...rest) -----
def total(*numbers):
    return sum(numbers)
print(total(1, 2, 3, 4))     # Output: 10

# ----- 6. **kwargs -> variable keyword args (TS: options object) -----
def describe(**info):
    return info
print(describe(name="Vetri", role="dev"))
# Output: {'name': 'Vetri', 'role': 'dev'}

# ----- 7. Returning multiple values (as a tuple) -----
def min_max(nums):
    return min(nums), max(nums)
low, high = min_max([4, 1, 9, 2])
print(low, high)             # Output: 1 9

# ----- 8. Lambda (anonymous function; TS: arrow function) -----
square = lambda x: x * x
print(square(6))             # Output: 36
# Common use: as a key/callback
nums = [3, 1, 2]
print(sorted(nums, key=lambda x: -x))   # Output: [3, 2, 1]

# Cross-language note:
#   Python `def f(x): ...`        == TS `function f(x) { ... }`
#   Python `lambda x: x*x`        == TS `(x) => x*x`
#   Python returns a tuple for "multiple returns"; TS returns an array/object.

File: rust/07_functions.rs
// ============================================================
// Topic:  Functions
// Summary: fn, return types, implicit return, closures, Option<T>.
// Run:     rustc rust/07_functions.rs -o rust/07_functions && rust/07_functions
// ============================================================

// ----- 1. Basic function -----
fn greet(name: &str) -> String {
    format!("Hello, {}", name)
}

// ----- 2. Type annotations (mandatory in Rust) -----
fn add(a: i32, b: i32) -> i32 {
    return a + b; // explicit return
}

// ----- 3. Implicit return (no semicolon on last expression) -----
fn multiply(a: i32, b: i32) -> i32 {
    a * b // no semicolon = returned value
}

// ----- 4. Default parameters — Rust has no defaults; use multiple fns or Option -----
fn power(base: i32, exp: i32) -> i32 {
    base.pow(exp as u32)
}

// ----- 5. Returning multiple values via tuple -----
fn min_max(nums: &[i32]) -> (i32, i32) {
    let mut min = nums[0];
    let mut max = nums[0];
    for &n in nums {
        if n < min {
            min = n;
        }
        if n > max {
            max = n;
        }
    }
    (min, max)
}

// ----- 6. Option<T> — safe alternative to null -----
fn safe_div(a: i32, b: i32) -> Option<i32> {
    if b == 0 {
        None
    } else {
        Some(a / b)
    }
}

fn main() {
    println!("{}", greet("Vetri")); // Output: Hello, Vetri

    println!("{}", add(2, 3)); // Output: 5

    println!("{}", multiply(4, 5)); // Output: 20

    println!("{}", power(5, 2)); // Output: 25
    println!("{}", power(5, 3)); // Output: 125

    let (low, high) = min_max(&[4, 1, 9, 2]);
    println!("{} {}", low, high); // Output: 1 9

    println!("{:?}", safe_div(10, 2)); // Output: Some(5)
    println!("{:?}", safe_div(10, 0)); // Output: None

    // ----- 7. Closures (anonymous functions) -----
    let square = |x: i32| x * x;
    println!("{}", square(6)); // Output: 36

    // Closure as callback (sort by descending)
    let mut nums = vec![3, 1, 2];
    nums.sort_by(|a, b| b.cmp(a));
    println!("{:?}", nums); // Output: [3, 2, 1]

    // ----- 8. Higher-order function -----
    fn apply(f: fn(i32) -> i32, x: i32) -> i32 {
        f(x)
    }
    fn double(n: i32) -> i32 {
        n * 2
    }
    println!("{}", apply(double, 7)); // Output: 14

    // Cross-language note:
    //   Rust `fn f(x) -> T { ... }`  == Python `def f(x):`  == TS `function f(x): T`
    //   Rust closures `|x| x*x`     == Python `lambda x: x*x`  == TS `(x) => x*x`
    //   Rust returns tuples for multiple values; Python/TS use tuples/arrays/objects.
    //   Rust has no default params — use builder patterns or Option arguments instead.
    //   Option<T> replaces null: Some(value) or None.
}